processor/abci.go

```diff
@@ -690,6 +714,12 @@ func (app *App) OnCommit() (resp tmtypes.ResponseCommit) {
 	app.updateStats()
 	app.setBatchStats()
 
+	// if there is an approved protocol upgrade proposal and the current block height is later than the proposal's block height then take a snapshot and wait to be killed by the process manager
+	if app.protocolUpgradeService.TimeForUpgrade() {
+		app.snapshot.SnapshotNow(app.blockCtx)
```

Comment:
This snapshot hash needs to be set as `resp.Data`. Also if it happens to already be a snapshot block, we'll be double-snapshotting and will have two snapshots for the same block height in the snapshotdb.

Reply:
re double snapshots - I'm not sure how the snapshotdb will handle that, but the reason I force a snapshot here regardless if one was taken or not is because of the asynchronous save of the snapshot which I can't rely on here.

Regarding to setting the snapshot hash in resp.Data, why do we need this? I don't want to force using it for consensus, I just want to make sure it's saved.

Reply:
ok